### PR TITLE
feat: add android-sdk-framework module

### PR DESCRIPTION
**Pull request by typotter**
## Summary

Add new android-sdk-framework module that provides base client functionality and configuration storage abstractions for Android SDK implementations.

## Key Components

- **AndroidBaseClient**: Base class for Android SDK clients that extends the v4 Java SDK framework
- **CachingConfigurationStore**: Abstract store for cached configurations with lifecycle management
- **FileBackedConfigStore**: File-based implementation for configuration persistence
- **ConfigurationCodec**: Pluggable serialization/deserialization interface with default Java serialization
- **ByteStore**: Low-level byte storage abstraction
- **ConfigCacheFile**: Disk cache file management with content-type based extensions

## Dependencies

This module depends on:
- v4 Java SDK framework (cloud.eppo:eppo-sdk-framework)
- slf4j-android for logging

## Testing

- Framework module builds independently: ✅
- Framework unit tests pass: ✅
- Existing eppo module unchanged: ✅
- Existing eppo tests pass: ✅

## Notes

- This PR adds the framework module as a standalone component
- The eppo module is NOT modified and remains on v3 dependencies
- Integration with the eppo module will happen in a follow-up PR
- Supersedes #248

## Summary by CodeRabbit

* **New Features**
  * Android SDK framework module with async initialization and flexible configuration options
  * Configuration persistence and file-backed storage capabilities
  * Pause and resume polling functionality for managing network updates

* **Documentation**
  * Added comprehensive development guide covering SDK structure, build commands, and publishing workflow

* **Tests**
  * Added instrumentation tests for polling behavior and pause/resume cycles
  * Added unit tests for storage components with concurrency and error handling validation

### CODE REVIEW

**Comment by typotter** [comment]
@copilot review

File: android-sdk-framework/src/androidTest/java/cloud/eppo/android/framework/EppoClientPollingTest.java [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+        .buildAndInitAsync()
+        .get();
+  }
+
+  /**
+   * Builds a client in offline mode without polling enabled.
+   *
+   * @return Initialized EppoClient
+   */
+  private AndroidBaseClient<JsonNode> buildOfflineClientWithoutPolling()
```
Comment (aarsilv) [inline android-sdk-framework/src/androidTest/java/cloud/eppo/android/framework/EppoClientPollingTest.java]: pretty much same as previous method but with a different boolean passed for `pollingEnabled()` and the call to `pollingIntervalMs()` Could be consolidated to a single method (with 0 pollingMs meaning no polling) or the explicitly named methods with a shared base method.

File: android-sdk-framework/src/androidTest/java/cloud/eppo/android/framework/EppoClientPollingTest.java [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+        .offlineMode(true)
+        .initialConfiguration(initialConfig)
+        .pollingEnabled(false)
+        .isGracefulMode(true) // Enable graceful mode to handle initialization issues
+        .buildAndInitAsync()
+        .get();
+  }
+
+  @Test
+  public void testPauseAndResumePolling() throws ExecutionException, InterruptedException {
```
Comment (aarsilv) [inline android-sdk-framework/src/androidTest/java/cloud/eppo/android/framework/EppoClientPollingTest.java:88]: This test doesn't seem to be doing much. Consider a shorter polling interval and then pausing longer than the interval to make sure no polls happens and then check polling happened again after unpausing.

File: android-sdk-framework/src/androidTest/java/cloud/eppo/android/framework/EppoClientPollingTest.java [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+    androidBaseClient.resumePolling();
+    Log.d(TAG, "Third resume");
+    Thread.sleep(50);
+
+    // Final cleanup
+    androidBaseClient.pausePolling();
+  }
+
+  @Test
+  public void testPauseResumeSequenceDoesNotCrash()
```
Comment (aarsilv) [inline android-sdk-framework/src/androidTest/java/cloud/eppo/android/framework/EppoClientPollingTest.java]: 👍

File: android-sdk-framework/src/androidTest/java/cloud/eppo/android/framework/EppoClientPollingTest.java [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+    Log.d(TAG, "Resume called without starting - should log warning");
+
+    // Wait a bit to ensure no crashes
+    Thread.sleep(50);
+
+    // Should not crash or throw exception
+  }
+
+  @Test
+  public void testMultiplePauseResumeCycles() throws ExecutionException, InterruptedException {
```
Comment (aarsilv) [inline android-sdk-framework/src/androidTest/java/cloud/eppo/android/framework/EppoClientPollingTest.java]: Seems like overkill

File: android-sdk-framework/src/main/java/cloud/eppo/android/framework/storage/FileBackedConfigStore.java [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+      @NotNull String cacheFileSuffix,
+      @NotNull ConfigurationCodec<Configuration> codec) {
+    super(codec, createByteStore(application, cacheFileSuffix, codec));
+  }
+
+  private static ByteStore createByteStore(
+      Application application, String cacheFileSuffix, ConfigurationCodec<Configuration> codec) {
+    ConfigCacheFile cacheFile =
+        new ConfigCacheFile(application, cacheFileSuffix, codec.getContentType());
+    return new FileBackedByteStore(cacheFile);
```
Comment (aarsilv) [inline android-sdk-framework/src/main/java/cloud/eppo/android/framework/storage/FileBackedConfigStore.java]: Pretty small--any reason not to inline it int he constructor?

File: android-sdk-framework/src/main/java/cloud/eppo/android/framework/AndroidBaseClient.java [hunk 6]
```diff
@@ -0,0 +1,425 @@
...
+
+/**
+ * Generic EppoClient that extends BaseEppoClient with JSON type parameter.
+ *
+ * <p>Requires callers to provide implementations of ConfigurationParser and
+ * EppoConfigurationClient.
+ *
+ * @param <JsonFlagType> The JSON type used for JSON flag values (e.g., JsonNode, JsonElement)
+ */
+public class AndroidBaseClient<JsonFlagType> extends BaseEppoClient<JsonFlagType> {
```
Comment (aarsilv) [inline android-sdk-framework/src/main/java/cloud/eppo/android/framework/AndroidBaseClient.java]: we seem to be inconsistent with where we put "base" in our name. Should this be `BaseAndroidClient` or `BaseEppoAndroidClient`?

And since this isn't extended anywhere, should it just be `EppoClient`? or `EppoAndroidClient`?

File: android-sdk-framework/src/main/java/cloud/eppo/android/framework/AndroidBaseClient.java [hunk 7]
```diff
@@ -0,0 +1,425 @@
...
+                            "Unable to initialize client; Configuration could not be loaded", ex));
+                  }
+                  return null;
+                });
+      }
+
+      // Start polling if configured
+      if (pollingEnabled && pollingIntervalMs > 0) {
+        Log.i(TAG, "Starting poller");
+        long effectiveJitter = pollingJitterMs;
```
Comment (aarsilv) [inline android-sdk-framework/src/main/java/cloud/eppo/android/framework/AndroidBaseClient.java]: consider adding units here in the name, `effectiveJitterMs`

File: android-sdk-framework/src/test/java/cloud/eppo/android/framework/storage/CachingConfigurationStoreTest.java [hunk 8]
```diff
@@ -0,0 +1,359 @@
...
+    } catch (ExecutionException e) {
+      assertTrue("Should contain RuntimeException", e.getCause() instanceof RuntimeException);
+      assertTrue("Should contain error message", e.getCause().getMessage().contains("Read failed"));
+    } catch (Exception e) {
+      fail("Unexpected exception: " + e.getMessage());
+    }
+  }
+
+  @Test
+  public void testThreadSafety_concurrentSaves() throws Exception {
```
Comment (aarsilv) [inline android-sdk-framework/src/test/java/cloud/eppo/android/framework/storage/CachingConfigurationStoreTest.java:251]: 💪

File: android-sdk-framework/src/test/java/cloud/eppo/android/framework/storage/ConfigurationCodecTest.java [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+  public void roundTrip_serializeAndDeserialize_succeeds() {
+    SerializableEppoConfiguration original =
+        (SerializableEppoConfiguration) Configuration.emptyConfig();
+    byte[] bytes = codec.toBytes(original);
+    assertNotNull(bytes);
+    assertTrue(bytes.length > 0);
+
+    SerializableEppoConfiguration deserialized = codec.fromBytes(bytes);
+    assertNotNull(deserialized);
+    assertEquals("Deserialized should equal original", original, deserialized);
```
Comment (aarsilv) [inline android-sdk-framework/src/test/java/cloud/eppo/android/framework/storage/ConfigurationCodecTest.java:109]: 👍

File: android-sdk-framework/src/test/java/cloud/eppo/android/framework/storage/FileBackedByteStoreTest.java [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+    byte[] originalData = "Test data for round trip".getBytes(StandardCharsets.UTF_8);
+
+    // Write
+    byteStore.write(originalData).get(5, TimeUnit.SECONDS);
+
+    // Read
+    byte[] readData = byteStore.read().get(5, TimeUnit.SECONDS);
+
+    assertNotNull("Read data should not be null", readData);
+    assertArrayEquals("Round trip should preserve data", originalData, readData);
```
Comment (aarsilv) [inline android-sdk-framework/src/test/java/cloud/eppo/android/framework/storage/FileBackedByteStoreTest.java:84]: 👍

File: android-sdk-framework/src/test/java/cloud/eppo/android/framework/storage/FileBackedByteStoreTest.java [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+    // Delete file
+    cacheFile.delete();
+
+    // Read after delete should return null
+    byte[] read2 = byteStore.read().get(5, TimeUnit.SECONDS);
+    assertNull("Read after delete should return null", read2);
+  }
+
+  @Test
+  public void testConcurrentWrites() throws Exception {
```
Comment (aarsilv) [inline android-sdk-framework/src/test/java/cloud/eppo/android/framework/storage/FileBackedByteStoreTest.java:173]: 💪

File: android-sdk-framework/build.gradle [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+
+    testOptions {
+        unitTests.returnDefaultValues = true
+    }
+}
+
+dependencies {
+    api 'cloud.eppo:eppo-sdk-framework:0.1.0-SNAPSHOT'
+
+    api 'com.google.code.gson:gson:2.10.1'
```
Comment (aarsilv) [inline android-sdk-framework/build.gradle]: I think this is now a test dependency only (isn't that the whole point of all of this 😅 )

File: android-sdk-framework/src/main/java/cloud/eppo/android/framework/BaseAndroidEppoClient.java [hunk 13]
```diff
@@ -0,0 +1,425 @@
...
+      }
+
+      if (newInstance.getInitialConfigFuture() != null) {
+        newInstance
+            .getInitialConfigFuture()
+            .handle(
+                (success, ex) -> {
+                  if (ex == null && Boolean.TRUE.equals(success)) {
+                    ret.complete(newInstance);
+                  } else if (offlineMode || failCount.incrementAndGet() == 2) {
```
Comment (aarsilv) [inline android-sdk-framework/src/main/java/cloud/eppo/android/framework/BaseAndroidEppoClient.java:373]: Had Claude give this a look and it is worried that `buildAndInitAsync()` can hang forever. 

```
  buildAndInitAsync() can hang forever. If initialConfigFuture is non-null and  
  resolves with success = false/null, the else branch increments    
  failCount twice (to 2, then 3). If loadConfigurationAsync then also fails,    
  incrementAndGet() returns 4 — the == 2 check never matches, and ret is never
  completed. The caller blocks forever on .get().
```

Also on that note, consider providing an optional timeout (with a default to something generous like 30 seconds?)

File: android-sdk-framework/src/main/java/cloud/eppo/android/framework/storage/BaseCacheFile.java [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+  /** Useful for passing in as a reader for JSON deserialization. */
+  public BufferedReader getReader() throws IOException {
+    return new BufferedReader(new FileReader(cacheFile));
+  }
+
+  /** Useful for mocking caches in automated tests. */
+  public void setContents(String contents) {
+    delete();
+    try (BufferedWriter writer = getWriter()) {
+      writer.write(contents);
```
Comment (aarsilv) [inline android-sdk-framework/src/main/java/cloud/eppo/android/framework/storage/BaseCacheFile.java:66]: Technically, a crash between these two will result in the loss of file. So you could do write and then rename. But if this is is only used for tests not worth worrying about this edge case.

File: android-sdk-framework/src/main/java/cloud/eppo/android/framework/BaseAndroidEppoClient.java (hunk 13 above)
Comment (aarsilv) [reply android-sdk-framework/src/main/java/cloud/eppo/android/framework/BaseAndroidEppoClient.java:373]: It also has a possible other recommendation regarding use of `anyOf()`

```
  ret.complete() can be called from two different callbacks on success. Both    
  loadConfigurationAsync and initialConfigFuture handlers call
  ret.complete(newInstance) on success. CompletableFuture.complete() is         
  idempotent so this isn't a bug, but it suggests the dual-callback design is
  more complex than it needs to be. A clearer approach would be
  CompletableFuture.anyOf() for success with an explicit failure combiner.
```

File: android-sdk-framework/src/main/java/cloud/eppo/android/framework/AndroidBaseClient.java [hunk 15]
```diff
@@ -0,0 +1,425 @@
...
+  private static final String TAG = logTag(AndroidBaseClient.class);
+  private static final boolean DEFAULT_IS_GRACEFUL_MODE = true;
+  private static final boolean DEFAULT_OBFUSCATE_CONFIG = true;
+  private static final long DEFAULT_POLLING_INTERVAL_MS = 5 * 60 * 1000;
+  private static final long DEFAULT_JITTER_INTERVAL_RATIO = 10;
+
+  private long pollingIntervalMs;
+  private long pollingJitterMs;
+
+  @Nullable private static AndroidBaseClient<?> instance;
```
Comment (aarsilv) [inline android-sdk-framework/src/main/java/cloud/eppo/android/framework/AndroidBaseClient.java]: By not having a default no-op instance, `buildAndInit()` _could_ return null if graceful mode is at play and an error happens before instance is set.

File: android-sdk-framework/build.gradle [hunk 16]
```diff
@@ -0,0 +1,156 @@
...
+    api 'org.slf4j:slf4j-android:1.7.36'
```
Comment (aarsilv) [inline android-sdk-framework/build.gradle]: Don't see this in the code, should this be `runtimeOnly`?

File: android-sdk-framework/src/main/java/cloud/eppo/android/framework/storage/CachingConfigurationStore.java [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+    if (config == null) {
+      throw new IllegalArgumentException("config must not be null");
+    }
+    byte[] bytes = codec.toBytes(config);
+    return byteStore
+        .write(bytes)
+        .thenRun(
+            () -> {
+              this.configuration = config;
+            });
```
Comment (aarsilv) [inline android-sdk-framework/src/main/java/cloud/eppo/android/framework/storage/CachingConfigurationStore.java]: Had Claude dig into this, and it thinks there a race condition.

```
  The write goes to the single-thread IO_EXECUTOR, but thenRun executes on      
  whatever thread completes the future — typically the IO thread, but not
  guaranteed. Here's the race:                                                  
                  
  Thread 1: saveConfiguration(A)
    1. codec.toBytes(A)                                                         
    2. byteStore.write(A) → queued on IO_EXECUTOR
                                                                                
  Thread 2: saveConfiguration(B)
    3. codec.toBytes(B)                                                         
    4. byteStore.write(B) → queued on IO_EXECUTOR
                                                                                
  IO_EXECUTOR runs them in order:
    5. writes A to disk                                                         
    6. writes B to disk       ← disk now has B                                  
   
  thenRun callbacks fire:                                                       
    7. this.configuration = B ← from step 4's callback
    8. this.configuration = A ← from step 2's callback (ran slightly later)     
                                                                                
  Result: disk=B, memory=A
                                                                                
  Step 8 can run after step 7 because thenRun doesn't run on the IO_EXECUTOR —  
  it runs on whatever thread calls complete(), or the calling thread if the
  future is already done. The two thenRun callbacks are not ordered by the      
  single-thread executor.

  The simplest fix is to run the in-memory update inside the IO_EXECUTOR too:   
   
  return byteStore                                                              
      .write(bytes)
      .thenRunAsync(() -> { this.configuration = config; }, IO_EXECUTOR);
                                                                                
  This guarantees the disk write and memory update for A both happen before B's,
   since they're all serialized on the same single thread. Though that would    
  require exposing the executor or moving the logic into FileBackedByteStore.
```

**Review by aarsilv — APPROVED** [review]
Nice work! You've really put together a flexible, extensible, and well-tested Android SDK with many lessons we should take with us to Datadog. (On that note, please write up lessons from all this in a confluence page). I like the clean code and good test coverage including concurrency tests.

In addition to my own review, I added co-pilot and also had Claude take a pass. I'm approving as-is, although there are some things that I and the tools found that I'd love for you to engage on: 

- EppoClientPollingTest should check for actual poll calls (or lack of, when paused) using spys or some other mechanism
- Copilot-identified leaky newSingleThreadExecutor
- Claude-identified saveConfiguration disk/memory race condition
- Name of AndroidBaseClient
- Consider a default no-op instance
- Claude-identified possibility of `buildAndInit()` hanging forever
- Copilot-identified need of atomic singleton and/or thread safe access of instance in AndroidBaseClient (volatile/synchronized)
- gson shouldn't be a dependency anymore
- slf4j-android maybe should be runtime only

**Comment by aarsilv** [comment]
I asked Claude to check past PRs for past issues and make sure we're not reintroducing them, it came up:
```
  1. "The async cache load race from PR #235 appears to be re-introduced —      
  loadFromStorage() doesn't update the in-memory cache, so getConfiguration() 
  returns empty until the first saveConfiguration()"                            
  2. "The SNAPSHOT dependency in api scope will break consumers if published 
  as-is (same issue as #78)"    
```

Details:
```
  1. Async cache load race (PR #235 redux)                                      
                                                                                
  In buildAndInitAsync() (line ~672-674):                                       
   
  if (initialConfiguration == null && !ignoreCachedConfiguration) {             
      initialConfiguration = configStore.loadFromStorage();                     
  }
                                                                                
  loadFromStorage() reads from disk and returns the config via a                
  CompletableFuture, but deliberately does not update the in-memory cache (line
  ~1022, the Javadoc even says "without updating the in-memory cache"). That    
  loaded config gets passed to the BaseEppoClient constructor as
  initialConfiguration.

  Meanwhile, configStore.getConfiguration() still returns emptyConfig() — it    
  won't change until someone calls saveConfiguration(), which only happens when
  a fresh config is fetched from the network.                                   
                  
  So in this sequence:                                                          
  1. App starts, cache file exists on disk with valid config
  2. buildAndInitAsync() loads cache into initialConfiguration future           
  3. Client is constructed, configStore.getConfiguration() returns empty
  4. User calls a flag evaluation that goes through                             
  configStore.getConfiguration() → gets empty config → returns default value    
  5. Network fetch completes, calls saveConfiguration() → now in-memory cache is
   populated                                                                    
                                                                                
  This is the exact scenario from PR #235: offline or slow-network users get
  empty config despite having a perfectly valid cache on disk. The fix from PR  
  #235 was to update this.configuration inside the async load — that same fix is
   missing here.                                                                
                  
  ---                                                                           
  2. SNAPSHOT dependency in api scope
                                                                                
  In build.gradle (line 58):

  api 'cloud.eppo:eppo-sdk-framework:0.1.0-SNAPSHOT'
                                                                                
  Two problems:
                                                                                
  Scope: api means this dependency is transitive — any app that depends on      
  android-sdk-framework will need to resolve eppo-sdk-framework:0.1.0-SNAPSHOT
  at compile time. SNAPSHOT artifacts are only available from a snapshot        
  repository (typically Sonatype OSSRH snapshots), which most consumers don't
  have configured. This is what caused Issue #78 when
  sdk-common-jvm:2.0.0-SNAPSHOT leaked to consumers in v3.2.0.

  Timing: If this gets published to Maven Central as-is (even accidentally, or  
  as a pre-release), consumers' builds will fail with a dependency resolution
  error. The 0.1.0-SNAPSHOT needs to be promoted to a release version before    
  this module is published — but there's no build guard preventing a premature
  publish. The shouldPublish flag in the Gradle config is a manual check, not an
   automated one.

  Suggestion: add a build-time check that fails the publish task if any         
  api/implementation dependency contains SNAPSHOT.

```

File: android-sdk-framework/src/main/java/cloud/eppo/android/framework/AndroidBaseClient.java (hunk 15 above)
Comment (aarsilv) [reply android-sdk-framework/src/main/java/cloud/eppo/android/framework/AndroidBaseClient.java]: Also, this is to be extended by EppoClient which has its own singleton so I don't think we want this one here nor does it make sense to have multiple singleton instances throughout the inheritance tree.

File: android-sdk-framework/src/main/java/cloud/eppo/android/framework/AndroidBaseClient.java (hunk 6 above)
Comment (aarsilv) [reply android-sdk-framework/src/main/java/cloud/eppo/android/framework/AndroidBaseClient.java]: Still stand by a rename, such as `AndroidEppoClient`

File: android-sdk-framework/build.gradle [hunk 18]
```diff
@@ -0,0 +1,181 @@
...
+task checkVersion {
+    dependsOn checkNoSnapshotApiDeps
+    doLast {
+        if (!project.hasProperty('release') && !project.hasProperty('snapshot')) {
+            throw new GradleException("You must specify either -Prelease or -Psnapshot")
+        }
+        if (project.hasProperty('release') && project.version.endsWith('SNAPSHOT')) {
+            throw new GradleException("You cannot specify -Prelease with a SNAPSHOT version")
+        }
+        if (project.hasProperty('snapshot') && !project.version.endsWith('SNAPSHOT')) {
```
Comment (aarsilv) [inline android-sdk-framework/build.gradle:162]: 🤖 Comment from AI: This is the line that makes the snapshot publish path impossible for the new module. Since `version` is fixed at `0.1.0`, `./gradlew :android-sdk-framework:publish -Psnapshot` always throws here (`You cannot specify -Psnapshot with a non-SNAPSHOT version`). Release publish is blocked too while the `api` dependency is still `0.1.0-SNAPSHOT`. Could we either make the module version `0.1.0-SNAPSHOT` until upstream is released, or derive the version from the `snapshot`/`release` flag so the snapshot flow can actually publish?

File: android-sdk-framework/src/main/java/cloud/eppo/android/framework/storage/FileBackedByteStore.java [hunk 19]
```diff
@@ -0,0 +1,88 @@
...
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * {@link ByteStore} implementation that reads and writes a single file via {@link BaseCacheFile}.
+ *
+ * <p>Implements {@link java.io.Closeable}. Call {@link #close()} when the store is no longer needed
+ * to release the background IO thread. If not closed explicitly, the daemon thread will be
+ * reclaimed by the JVM/Android runtime on process exit.
+ */
+public final class FileBackedByteStore implements ByteStore, java.io.Closeable {
```
Comment (aarsilv) [inline android-sdk-framework/src/main/java/cloud/eppo/android/framework/storage/FileBackedByteStore.java:15]: 🤖 Comment from Claude: `close()` was added so the single-thread `ioExecutor` can be reclaimed, but nothing in the production path ever calls it — neither `CachingConfigurationStore` nor `FileBackedConfigStore` is `Closeable`, and `BaseAndroidClient` holds the store without closing it. For a long-lived singleton store this is fine (the thread is a daemon). But `Builder.forceReinitialize(true)` constructs a brand-new `FileBackedConfigStore` → new `FileBackedByteStore` → new executor thread each time, with no path to shut down the previous one. Repeated reinitialization leaks one daemon thread per cycle. Consider plumbing `close()` through `CachingConfigurationStore`/`FileBackedConfigStore` and closing the prior store on reinitialize.

File: android-sdk-framework/src/main/java/cloud/eppo/android/framework/util/Utils.java [hunk 20]
```diff
@@ -0,0 +1,31 @@
...
+    if (key == null || key.isEmpty()) {
+      return "";
+    }
+    // Take the first eight characters to avoid the key being sensitive information.
+    // Remove non-alphanumeric characters so it plays nice with filesystem paths.
+    // \W is equivalent to [^a-zA-Z0-9_] — underscores are kept, which is fine for filenames.
+    // Note: if the first 8 characters are all non-word the result is an empty string,
+    // which produces the filename "eppo-sdk-flags-.bin". Eppo-issued API keys always start
+    // with alphanumeric characters, so this edge case is not expected in production.
+    return key.substring(0, Math.min(8, key.length())).replaceAll("\\W", "");
```
Comment (aarsilv) [inline android-sdk-framework/src/main/java/cloud/eppo/android/framework/util/Utils.java:31]: 🤖 Comment from Claude: **Updated — lower priority after checking the backend.** I traced SDK-key generation in the Eppo monorepo (`api-key.service.ts`). Keys are `{randomBytes}.{encodedPayload}`: the high-entropy random portion (≥8 chars, `MIN_RANDOM_BYTES = 8`) comes **before** the period, and the encoded domain/routing info comes **after**. The backend itself keys off `token.substr(0, 8)` as a uniqueness prefix, so the first 8 chars are effectively unique per key — collision risk for real Eppo keys is negligible.

Two small leftover notes, take them or leave them:
1. `\\W` strips base64url `-` (but keeps `_`), so a `-` among those first 8 chars slightly shortens the effective key. Negligible in practice.
2. If a non-Eppo / custom key (no random prefix) is ever passed, the 8-char-prefix assumption breaks. A one-line comment noting the dependency on Eppo's key format would be enough.

File: android-sdk-framework/src/main/java/cloud/eppo/android/framework/storage/CachingConfigurationStore.java [hunk 21]
```diff
@@ -0,0 +1,132 @@
...
+   * is found.
+   *
+   * <p>Combines {@link #loadFromStorage()} and {@link #seedCache(Configuration)} so callers do not
+   * need to reach into the storage package to call the package-private {@code seedCache} method
+   * directly.
+   *
+   * @return a future that completes with the loaded configuration, or null if storage is empty or
+   *     missing
+   */
+  @NotNull public CompletableFuture<Configuration> loadAndSeedFromStorage() {
```
Comment (aarsilv) [inline android-sdk-framework/src/main/java/cloud/eppo/android/framework/storage/CachingConfigurationStore.java:118]: 🤖 Comment from Claude: Minor efficiency observation: `loadAndSeedFromStorage()` both seeds the in-memory cache *and* returns the loaded config, which `Builder.buildAndInitAsync()` then passes as `initialConfiguration`. Per the deadlock-fix commit (8d16eeb), the framework's `ConfigurationRequestor.setInitialConfiguration()` chains into `saveConfiguration()` → `byteStore.write()`. So on every cold start with a valid cache, the SDK reads the cached bytes and then immediately writes the identical bytes back to disk. Functionally harmless, but it's an avoidable disk write on each launch. Seeding the in-memory cache (which you already do) should be sufficient to make `getConfiguration()` return immediately without re-persisting.

File: settings.gradle [hunk 22]
```diff
@@ -19,3 +19,4 @@ dependencyResolutionManagement {
 rootProject.name = "Eppo SDK"
 include ':example'
 include ':eppo'
+include ':android-sdk-framework'
```
Comment (aarsilv) [inline settings.gradle:22]: 🤖 Comment from Claude: Nit: no trailing newline at end of file. The module's spotless `misc` block enforces `endWithNewline()` on `*.gradle`, but it's scoped to the `android-sdk-framework` dir so this root file isn't caught by CI — still worth adding the newline for consistency.

File: android-sdk-framework/src/main/java/cloud/eppo/android/framework/util/Utils.java (hunk 20 above)
Comment (aarsilv) [reply android-sdk-framework/src/main/java/cloud/eppo/android/framework/util/Utils.java:31]: This comment can be safely ignored, no practical implications

**Comment by typotter** [comment]
Superseded by #260 (2-param generics)
